imagebuildah/build.go

```diff
@@ -448,13 +448,53 @@ func (b *Executor) volumeCacheRestore() error {
 // Copy copies data into the working tree.  The "Download" field is how
 // imagebuilder tells us the instruction was "ADD" and not "COPY".
 func (b *Executor) Copy(excludes []string, copies ...imagebuilder.Copy) error {
+	excludes, err := imagebuilder.ParseDockerignore(b.contextDir)
+	if err != nil {
+		return err
+	}
 	for _, copy := range copies {
 		logrus.Debugf("COPY %#v, %#v", excludes, copy)
 		if err := b.volumeCacheInvalidate(copy.Dest); err != nil {
 			return err
 		}
 		sources := []string{}
 		for _, src := range copy.Src {
+			srcInfo, err := os.Stat(src)
+			if err != nil {
+				return err
+			}
+			excludeSrc := false
+			if !srcInfo.IsDir() {
+				for _, p := range excludes {
+					if match, err := filepath.Match(p, filepath.Clean(src)); match {
+						excludeSrc = true
+						break
+					} else if err != nil {
+						return err
+					}
+				}
+			} else {
+				err := filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
+					if err != nil {
+						return err
+					}
+					for _, p := range excludes {
+						if match, err := filepath.Match(p, path); match {
```

Comment:
should this logic be pushed to `Builder.Add` where we can check each single entry not the src?